datastream/client.go

```diff
@@ -0,0 +1,696 @@
+	case <-ctx.Done():
+		err = ctx.Err()
+	}
+
+	// Clean up all channels involved in this request
+	c.cleanupPendingCompanyRequests(cacheKeys, waitCh)
+
+	return nil, err
+}
+
+func (c *DataStreamClient) getUser(ctx context.Context, keys map[string]string) (*rulesengine.User, error) {
+	user := c.getUserFromCache(keys)
+	if user != nil {
+		return user, nil
+	}
+
+	waitCh := make(chan *rulesengine.User, 1) // Register the wait channel for each key
+	cacheKeys := []string{}
+	c.mu.Lock()
+	shouldSendRequest := true
+	for key, value := range keys {
+		cacheKey := resourceKeyToCacheKey(cacheKeyPrefixUser, key, value) // If there are already pending requests for this key, just add our channel
+		cacheKeys = append(cacheKeys, cacheKey)
+		if existingChannels, ok := c.pendingUserRequests[cacheKey]; ok {
+			c.pendingUserRequests[cacheKey] = append(existingChannels, waitCh)
+			shouldSendRequest = false // Someone else already requested this
+		}
+	}
+	c.mu.Unlock()
+
+	if shouldSendRequest {
+		req := &DataStreamReq{
+			EntityType: EntityTypeUser,
+			Keys:       keys,
+		}
+
+		err := c.SendWebSocketMessage(ctx, req)
+		if err != nil {
+			return nil, err
+		}
+
+		c.mu.Lock()
+		for key, value := range keys {
+			cacheKey := resourceKeyToCacheKey(cacheKeyPrefixUser, key, value)
+			c.pendingUserRequests[cacheKey] = []chan *rulesengine.User{waitCh}
+		}
+		c.mu.Unlock()
+	}
+
+	var err error
+	select {
+	case user := <-waitCh:
+		return user, nil
+	case <-time.After(resourceTimeout):
+		err = fmt.Errorf("timeout while waiting for user data")
+	case <-ctx.Done():
+		err = ctx.Err()
+	}
+
+	// Clean up all channels involved in this request
+	c.cleanupPendingUserRequests(cacheKeys, waitCh)
+
+	return nil, err
+}
+
+func (c *DataStreamClient) getFlag(ctx context.Context, key string) (*rulesengine.Flag, bool) {
+	flag, exists := c.flagsCacheProvider.Get(ctx, flagCacheKey(key))
+	if !exists {
+		return nil, false
+	}
+	return flag, true
+}
+
+func (c *DataStreamClient) handlePong(string) error {
+	err := c.conn.SetReadDeadline(time.Now().Add(pongWait))
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *DataStreamClient) packageMessage(req *DataStreamReq) *DataStreamBaseReq {
+	return &DataStreamBaseReq{
+		Data: *req,
+	}
+}
+
+func (c *DataStreamClient) getCompanyFromCache(keys map[string]string) *rulesengine.Company {
+	for key, value := range keys {
+		companyKey := resourceKeyToCacheKey(cacheKeyPrefixCompany, key, value)
+		company, exists := c.companyCacheProvider.Get(context.Background(), companyKey)
+		if exists {
+			return company
+		}
+	}
+
+	return nil
+}
+
+func (c *DataStreamClient) getUserFromCache(keys map[string]string) *rulesengine.User {
+	for key, value := range keys {
+		userKey := resourceKeyToCacheKey(cacheKeyPrefixUser, key, value)
+		user, exists := c.userCacheProvider.Get(context.Background(), userKey)
+		if exists {
+			return user
+		}
+	}
+
+	return nil
+}
+
+func flagCacheKey(key string) string {
```

Comment:
Do you think we should bust this cache if the library version changes? Kind of like what we do in our own app?

This way if a type changes, we're not storing a cached value that potentially can't be marshaled into the new type?